src/language-js/needs-parens.js

```diff
@@ -34,11 +34,37 @@ function hasClosureCompilerTypeCastComment(text, path, locStart, locEnd) {
         comment =>
           comment.leading &&
           comments.isBlockComment(comment) &&
-          comment.value.match(/^\*\s*@type\s*{[^}]+}\s*$/) &&
+          isTypeCastComment(comment.value) &&
           util.getNextNonSpaceNonCommentCharacter(text, comment, locEnd) === "("
       )
     );
   }
+
+  function isTypeCastComment(comment) {
+    const trimmed = comment.trim();
+    if (!/^\*\s*@type\s*\{/.test(trimmed)) {
+      return false;
+    }
+    const brackets = [];
```

Comment:
It'd be clearer this way:

```js
let isCompletelyClosed = false;
let unpairedBracketCount = 0;
for (const char of trimmed) {
  if (char === "{") {
    if (isCompletelyClosed) {
      return false;
    }
    unpairedBracketCount++;
  } else if (char === "}") {
    if (unpairedBracketCount === 0) {
      return false;
    }
    unpairedBracketCount--;
    if (unpairedBracketCount === 0) {
      isCompletelyClosed = true;
    }
  }
}
return unpairedBracketCount === 0;
```

Reply:
This logic doesn't require the {} to contain any content so technically `@type {}` is also valid. i guess in order to fully validate the syntax, we'd need doctrine or a mini parser for the syntax, which is probably unnecessary for this.